Add Checkbox 'Indeterminate' State styling

modialpad opened the pull request:
## Description

Add Checkbox 'Indeterminate' state styles. 

How it looks:

![image](https://user-images.githubusercontent.com/88498639/135866408-ec2f0256-96f4-43e1-b7c5-67798ec92be5.png)



## Pull Request Checklist

 - [x] Ask the contributors if a similar effort is already in process or has been solved.
 - [x] Review the [contribution guidelines](https://github.com/dialpad/dialtone/blob/staging/.github/CONTRIBUTING.md).
 - [x] Use `staging` as your pull request's base branch. (All PRs using `production` as its base branch will be declined).
 - [x] Ensure all `gulp` scripts successfully compile.
 - [x] Update, remove, or extend all affected documentation.
 - [x] Provide a description what is being changed, extended, or removed.
 - [x] Ensure no private Dialpad links or info are in the code or pull request description (Dialtone is a public repo!).
 - [x] Request a review from Joshua Hynes, Ted Goas, or David Becher.

dialpad-drew commented:
Awesome, @modialpad thanks for PRing this. Glad to see you updated the documentation.

I just pushed a quick update as well; my suggestion is we parse the indeterminate style out to its own discrete section. LMK what you think.

Also, out of curiosity: is your local build looping when you save?

modialpad commented:
> Awesome, @modialpad thanks for PRing this. Glad to see you updated the documentation.
> 
> I just pushed a quick update as well; my suggestion is we parse the indeterminate style out to its own discrete section. LMK what you think.
> 
> Also, out of curiosity: is your local build looping when you save?

Np, looks more organized now 👍 

Yes, when I run `npm run start` I think it rebuilds everything whenever I edit any files. But I think there's a bug where if anything changes the doc server doesn't restart properly. I had to exist the process and re-run `npm run start` in order to see app loading properly on `localhost:4000`. Is it just me?

dialpad-drew commented:
> > Awesome, @modialpad thanks for PRing this. Glad to see you updated the documentation.
> > I just pushed a quick update as well; my suggestion is we parse the indeterminate style out to its own discrete section. LMK what you think.
> > Also, out of curiosity: is your local build looping when you save?
> 
> Np, looks more organized now 👍
> 
> Yes, when I run `npm run start` I think it rebuilds everything whenever I edit any files. But I think there's a bug where if anything changes the doc server doesn't restart properly. I had to exist the process and re-run `npm run start` in order to see app loading properly on `localhost:4000`. Is it just me?

No, that's not just you. I'm noticing that on my end as well; that's new. I already noted it.

Only question I have is about the attribute. [MDN suggests](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/Input/checkbox) that browsers don't currently support `indeterminate` as an attribute and the example using the attribute doesn't seem to exhibit any functionality. Is there a CSS-only way to do this, or are we simply just putting in some styles for the Vue component to leverage?

dbecherdialpad commented on docs/components/checkbox.html:
```diff
@@ -38,75 +38,138 @@
+  </div>
+  <div class="d-checkbox-group">
+    <div class="d-checkbox__input">
+      <input class="d-checkbox" type="checkbox" name="Dialtone-CheckExample6" id="Dialtone-CheckExample6" indeterminate />
```
I don't think this is valid. There is no `indeterminate` attribute, you actually can ONLY set this through JS: https://css-tricks.com/indeterminate-checkboxes/

dbecherdialpad replied:
dammit @dialpad-drew already beat me to it

modialpad replied:
Done

dbecherdialpad commented on lib/build/less/components/radio-checkbox.less:
```diff
@@ -169,6 +169,18 @@
         }
+
+        &--indeterminate,
+        &[indeterminate] {
```
as above, this is not an attribute. But there IS a CSS pseudo-class weirdly (if it's set via js) https://developer.mozilla.org/en-US/docs/Web/CSS/:indeterminate

modialpad replied:
Done

dbecherdialpad commented on lib/build/less/components/radio-checkbox.less:
```diff
@@ -169,6 +169,18 @@
+            --check-radio--bgc:     var(--check-radio--color);
+            background-image: url("data:image/svg+xml,%3Csvg width='12' height='12' viewBox='0 0 24 24' xmlns='http://www.w3.org/2000/svg'%3E%3Cpath d='M19 13H5v-2h14v2z' fill='%23fff'/%3E%3C/svg%3E");
+
+            &[disabled] {
```
unrelated to your PR but I just noticed we aren't using the pseudo class for this either https://developer.mozilla.org/en-US/docs/Web/CSS/:disabled

modialpad replied:
Yeah, the other PR I have adds the custom class for the `disabled` state for accessibility reasons.

modialpad commented:
> Only question I have is about the attribute. [MDN suggests](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/Input/checkbox) that browsers don't currently support `indeterminate` as an attribute and the example using the attribute doesn't seem to exhibit any functionality. Is there a CSS-only way to do this, or are we simply just putting in some styles for the Vue component to leverage?

@dialpad-drew @dbecherdialpad I read the docs and you're right, browsers don't support `indeterminate` attribute on input elements. I removed the attribute selector and updated this select on the `indeterminate` pseudo-class, in addition to the custom class CSS approach (`--indeterminate`). Removed the attribute selector example in the docs. 


![image](https://user-images.githubusercontent.com/88498639/135638263-2c19914d-a52b-42e7-bf33-19d6be65a4d5.png)

modialpad commented:
@dbecherdialpad we probably don't need to update `dialtone-vue` for this, right? Considering that the consumer won't bother updating the `indeterminate` attribute through JS. And if it's just a class, then we can do that just through dialtone. 

Also, another question I have is about the publish process. I assume I'll have to bump the version of Dialtone after this merge, is there documentation somewhere for this?

(Finally, I can't seem to have merge permissions here. But we should probably go ahead and merge if it looks good)

dbecherdialpad commented:
@modialpad

> we probably don't need to update dialtone-vue for this, right? Considering that the consumer won't bother updating the indeterminate attribute through JS.

For dialtone-vue I generally prefer adding a prop (which adds the class) rather than relying on engineers to pass the correct class name. This also allows for improved autocomplete and documentation. For `indeterminate` we can also provide this as a prop and have the vue component correctly set the `indeterminate` state of the checkbox (essentially implementing what you thought could be done in html).

> Also, another question I have is about the publish process. I assume I'll have to bump the version of Dialtone after this merge, is there documentation somewhere for this?

Docs are here but it requires one of us to publish the version https://github.com/dialpad/dialtone/blob/staging/.github/RELEASING.md

dbecherdialpad commented:
@modialpad In dialtone-vue this is actually quite trivial to add using the `.prop` v-bind modifier: https://github.com/vuejs/vue/issues/4094#issuecomment-261659740

https://vuejs.org/v2/api/#v-bind

modialpad commented:
> @modialpad In dialtone-vue this is actually quite trivial to add using the `.prop` v-bind modifier: [vuejs/vue#4094 (comment)](https://github.com/vuejs/vue/issues/4094#issuecomment-261659740)
> 
> https://vuejs.org/v2/api/#v-bind

Great, thanks! I'll open the `dialtone-vue` PR once Dialtone is published.

juliodialpad reviewed (commented):
So this rule is going to raise a warning any time someone uses any of the mentioned properties?

This warning is going to be omnipresent on firespotter once enabled 😅 is this the validation we're looking for or should we be a bit more flexible here?

@braddialpad @francisrupert

ninamarina commented:
> So this rule is going to raise a warning any time someone uses any of the mentioned properties?
> 
> This warning is going to be omnipresent on firespotter once enabled 😅 is this the validation we're looking for or should we be a bit more flexible here?
> 
> @braddialpad @francisrupert

Totally, I would like to know what @francisrupert thinks about this one and the utility classes version as well: https://github.com/dialpad/dialtone/pull/459

braddialpad commented:
> So this rule is going to raise a warning any time someone uses any of the mentioned properties?

I think this is in fact the goal. It should be rare these properties are used directly.

braddialpad reviewed (approved):
Fine by me. It's obviously going to cause tons of warnings but I think being noisy is the intention.
